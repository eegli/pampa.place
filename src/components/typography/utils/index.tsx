import Box from '@mui/material/Box';

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
