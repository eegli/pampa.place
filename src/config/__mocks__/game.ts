import {GameConfig} from '../types';

export const config: GameConfig = {
  maxPlayers: 4,
  rounds: [1, 2, 3],
  roundsDefault: 2,
  timeLimits: [30, 60, 120, -1],
  timeLimitsDefault: 30,
};
