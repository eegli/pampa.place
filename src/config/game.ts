type GameConfig = {
  maxPlayers: number;
  roundSelect: [number, number, number];
  timeLimit: number;
};

const gameConfig: GameConfig = {
  maxPlayers: 4,
  roundSelect: [1, 3, 5],
  timeLimit: 10,
};

export default gameConfig;
