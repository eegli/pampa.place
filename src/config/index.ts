import { defaultSettings, DefaultSettings } from './defaults';
import customMaps from './maps';
import { markerColors } from './markers';

export type MarkerColors = Record<string, string>;

export type MapBounds = Record<
  'latMax' | 'latMin' | 'lngMax' | 'lngMin',
  number
>;
export type MapLatLng = Record<'lat' | 'lng', number>;
export type MapCenter = { center: MapLatLng };
export type MapEdges = Record<'NE' | 'SE' | 'SW' | 'NW', MapLatLng>;

export type UserMapConfig = MapBounds & { __name: string };

export type MapData = UserMapConfig & MapCenter & MapEdges;

interface Config {
  maps: MapData[];
  defaults: DefaultSettings;
  markers: string[];
}

const makeConfig = (): Config => {
  return {
    markers: Object.values(markerColors),
    maps: customMaps.reduce((acc, curr) => {
      acc.push({
        ...curr,
        center: {
          lat: (curr.latMax + curr.latMin) / 2,
          lng: (curr.latMax + curr.latMin) / 2,
        },
        NE: { lat: curr.latMax, lng: curr.lngMax },
        SE: { lat: curr.latMin, lng: curr.lngMax },
        SW: { lat: curr.latMin, lng: curr.lngMin },
        NW: { lat: curr.latMax, lng: curr.lngMin },
      });
      return acc;
    }, [] as MapData[]),
    defaults: defaultSettings,
  };
};

export default makeConfig();
