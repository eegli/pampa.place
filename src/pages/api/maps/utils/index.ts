export {apiData as data} from '@/config/maps';
import {NextApiHandler} from 'next';

type ErrorResponse = {
  error: string;
};

type DataResponse<T> = T extends unknown[]
  ? {
      data: T;
    }
  : {
      data: T[];
    };

export type ApiJSONHandler<T = unknown> = NextApiHandler<
  DataResponse<T> | ErrorResponse
>;
export type ApiTextHandler = NextApiHandler<string> | ErrorResponse;

export function nthQuery<T>(arg: T | T[], index = 0): T | undefined {
  return Array.isArray(arg) ? arg[index] : arg;
}
