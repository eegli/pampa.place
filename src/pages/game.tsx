import { NextPage } from 'next';
import { useRouter } from 'next/dist/client/router';
import { useEffect } from 'react';
import Error from '../components/error';
import Play from '../components/play/play';
import RoundEnd from '../components/round/round.end';
import RoundIntermission from '../components/round/round.intermission';
import RoundResult from '../components/round/round.result';
import { STATUS } from '../redux/game';
import { useAppDispatch, useAppSelector } from '../redux/hooks';
import { getRandomStreetView } from '../redux/position';
import { PageContentContainer } from '../styles';

const Game: NextPage = () => {
  const dispatch = useAppDispatch();
  const router = useRouter();

  const initialized = useAppSelector(({ game }) => game.initialized);
  const status = useAppSelector(({ game }) => game.status);
  const positionError = useAppSelector(({ position }) => position.error);

  useEffect(() => {
    if (!initialized) {
      router.push('/');
    }
  }, [initialized]);

  async function handleRetry() {
    await dispatch(getRandomStreetView({ radius: 1000 }));
  }

  function render() {
    switch (status) {
      case STATUS.INTERMISSION:
        return <RoundIntermission />;
      case STATUS.ROUND_STARTED:
        return <Play />;
      case STATUS.ROUND_ENDED:
        return <RoundEnd />;
      case STATUS.FINISHED:
        return <RoundResult />;

      default:
        return <div />;
    }
  }

  // If there is a header included, the page content container should
  // take the remaining height. The header is 48p by default
  /* 
   <GameHeader />
      <PageContentContainer height="calc(100% - 48px)">
        {render()}
      </PageContentContainer>
    </>
  
  */
  return (
    <>
      <PageContentContainer height="100%">
        {positionError ? (
          <Error
            callback={handleRetry}
            title="Error getting Street View data"
            info="This is likely because the map you chose is a little
                too small or has little Street View coverage."
            reason={positionError.message}
          />
        ) : (
          render()
        )}
      </PageContentContainer>
    </>
  );
};

export default Game;
