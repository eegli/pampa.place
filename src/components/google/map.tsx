import {svgMarker, svgMarkerColors} from '@/components/google/marker';
import {config} from '@/config/google';
import {LatLngLiteral, MapData} from '@/config/types';
import {Result} from '@/redux/game';
import {useAppDispatch} from '@/redux/hooks';
import {updateSelectedPosition} from '@/redux/position';
import {MapService, MarkerService, PolyLineService} from '@/services/google';
import {useEffect, useRef} from 'react';

export type GoogleMapProps = {
  map: MapData;
  mode?: 'preview' | 'play' | 'result';
  results?: Pick<Result, 'selected' | 'name'>[];
  initialPosition?: LatLngLiteral;
};

export const GoogleMap = ({
  map,
  mode,
  results,
  initialPosition,
}: GoogleMapProps) => {
  const dispatch = useAppDispatch();
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const unmount = MapService.mount(ref.current);
      /* Order in constructor is important! SW, NE  */
      const bounds = new google.maps.LatLngBounds(
        new google.maps.LatLng(map.properties.bbLiteral.SW),
        new google.maps.LatLng(map.properties.bbLiteral.NE)
      );
      google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
        MapService.map.fitBounds(bounds, 0);
      });

      return () => {
        unmount();
      };
    }
  }, [map]);

  useEffect(() => {
    if (mode === 'preview') {
      console.info('PREVIEW MODE MOUNT');
      MapService.map.setOptions({
        ...config.map,
        gestureHandling: 'none',
        mapTypeId: 'roadmap',
        mapTypeControl: false,
      });

      const features = MapService.map.data.addGeoJson(map);

      MapService.map.data.setStyle({
        fillColor: '#003d80',
        fillOpacity: 0.2,
        strokeWeight: 0.8,
      });
      return () => {
        console.info('PREVIEW MODE UNMOUNT');
        features.forEach(feat => {
          MapService.map.data.remove(feat);
        });
      };
    }
  }, [mode, map]);

  useEffect(() => {
    if (mode === 'play') {
      console.info('PLAY MODE MOUNT');
      MapService.map.setOptions(config.map);

      const marker = new google.maps.Marker();
      marker.setMap(MapService.map);
      marker.setDraggable(true);

      MarkerService.add(marker);

      // This Google Map event is not typed unfortunately
      const listener = MapService.map.addListener('click', (e: unknown) => {
        const {latLng} = e as {latLng: google.maps.LatLng};
        marker.setPosition(latLng);
        dispatch(
          updateSelectedPosition({lat: latLng.lat(), lng: latLng.lng()})
        );
      });
      return () => {
        console.info('PLAY MODE UNMOUNT');
        if (listener) listener.remove();
        // https://developers.google.com/maps/documentation/javascript/markers#remove
        MarkerService.clearAllItems();
      };
    }
  }, [mode, dispatch]);

  useEffect(() => {
    if (mode === 'result' && initialPosition) {
      console.info('RESULT MODE MOUNT');
      MapService.map.setOptions(config.map);

      // Add original location marker
      const originMarker = new google.maps.Marker();
      originMarker.setPosition(initialPosition);
      originMarker.setMap(MapService.map);

      MarkerService.add(originMarker);

      // Add marker for each result
      results &&
        results.forEach((p, idx) => {
          if (p.selected) {
            const playerMarker = new google.maps.Marker({
              position: p.selected,
              map: MapService.map,
              label: {
                text: p.name,
                color: 'white',
                className: 'map-marker',
              },
              icon: {
                path: svgMarker.path,
                fillColor: `#${svgMarkerColors[idx]}`,
                fillOpacity: 1,
                anchor: new google.maps.Point(
                  svgMarker.anchor[0],
                  svgMarker.anchor[1]
                ),
                strokeWeight: 0,
                scale: 1,
                labelOrigin: new google.maps.Point(15, 60),
              },
            });

            const polyLine = new google.maps.Polyline({
              path: [initialPosition, p.selected],
              map: MapService.map,
              geodesic: true,
              strokeColor: `#${svgMarkerColors[idx]}`,
              strokeOpacity: 1.0,
              strokeWeight: 4,
            });
            PolyLineService.add(polyLine);
            MarkerService.add(playerMarker);
          }
        });

      return () => {
        console.info('RESULT MODE UNMOUNT');
        MarkerService.clearAllItems();
        PolyLineService.clearAllItems();
      };
    }
  }, [mode, initialPosition, results]);

  return (
    <div
      ref={ref}
      style={{
        height: '100%',
      }}
    />
  );
};
