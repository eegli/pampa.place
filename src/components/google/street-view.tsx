import {config} from '@/config/google';
import {useAppSelector} from '@/redux/hooks';
import {StreetViewService} from '@/services/google';
import {Fade} from '@mui/material';
import {useEffect, useRef} from 'react';

type GoogleStreetViewProps = {
  staticPos?: boolean;
};

export const GoogleStreetView = ({
  staticPos = false,
}: GoogleStreetViewProps) => {
  const ref = useRef<HTMLDivElement>(null);
  const panoId = useAppSelector(({position}) => position.panoId);

  // Initialization
  useEffect(() => {
    if (ref.current) {
      const unmount = StreetViewService.mount(ref.current);
      return () => {
        unmount();
      };
    }
  }, []);

  useEffect(() => {
    StreetViewService.sv.setPano(panoId);
  }, [panoId]);

  useEffect(() => {
    StreetViewService.sv.setOptions({
      ...config.streetview,
      ...(staticPos && {
        clickToGo: false,
        disableDoubleClickZoom: true,
        disableDefaultUI: true,
      }),
    });
  }, [staticPos]);

  return (
    <Fade in timeout={500}>
      <div
        ref={ref}
        style={{
          height: '100%',
          width: '100%',
        }}
      />
    </Fade>
  );
};
