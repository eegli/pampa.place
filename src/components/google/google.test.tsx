import {config} from '@/config/game';
import {
  MapService,
  MarkerService,
  PolyLineService,
  StreetViewService,
} from '@/services/google';
import {testMap} from '@/tests/fixtures/map';
import {createMockStore, render, screen} from '@/tests/utils';
import {mocked} from 'jest-mock';
import {GoogleMap, GoogleMapProps} from './map';
import {svgMarkerColors} from './marker';
import {GoogleStreetView} from './street-view';
// Mock implementation for listeners. The handler will be caught and
// called with the event it would get from google.maps.Map's click
// event. Unfortunately, this event has no official types

// Store a reference to the events that are called when the map is
// mounted

// eslint-disable-next-line @typescript-eslint/ban-types
const events: {event: string; func: Function}[] = [];
const removeEventListener = jest.fn();
jest
  .spyOn(MapService.map, 'addListener')
  .mockImplementation((event, handler) => {
    const clickEvent = {latLng: {lat: () => 8, lng: () => 8}};
    const func = () => handler(clickEvent);
    events.push({event, func});
    return {remove: removeEventListener};
  });

jest
  .spyOn(google.maps.event, 'addListenerOnce')
  .mockImplementation((_, event, handler) => {
    events.push({event, func: handler});
    return {remove: removeEventListener};
  });

// @ts-expect-error - fake at least one element in the array to check
// the cleanup function for the feature
jest.spyOn(MapService.map.data, 'addGeoJson').mockReturnValue(['Feature 1']);

afterEach(() => {
  jest.clearAllMocks();
  events.length = 0;
});

const services = {
  gmap: mocked(MapService, true),
  gsv: mocked(StreetViewService, true),
  gmarkers: mocked(MarkerService, true),
  gpolylines: mocked(PolyLineService, true),
};

describe('Google, Map', () => {
  it('renders and has containers in document', () => {
    const mapMountSpy = jest.spyOn(MapService, 'mount');
    const {unmount} = render(<GoogleMap map={testMap} />);
    expect(screen.getByTestId('__GMAP__CONTAINER__')).toBeInTheDocument();
    expect(screen.getByTestId('__GMAP__')).toBeInTheDocument();
    expect(screen.getByTestId('__GMAP__')).toHaveStyle('height:100%');
    expect(mapMountSpy).toHaveBeenCalledTimes(1);
    expect(services.gmap.map.setOptions).not.toHaveBeenCalled();
    expect(events).toHaveLength(1);
    expect(events[0].event).toBe('idle');
    events[0].func();
    expect(services.gmap.map.fitBounds).toHaveBeenCalledTimes(1);
    unmount();
  });
  it('has preview mode', () => {
    const {unmount} = render(<GoogleMap map={testMap} mode="preview" />);
    expect(services.gmap.map.setOptions.mock.calls[0]).toMatchSnapshot(
      'preview settings'
    );
    expect(services.gmap.map.data.setStyle).toHaveBeenCalledTimes(1);
    expect(services.gmap.map.data.addGeoJson).toHaveBeenCalledTimes(1);
    expect(services.gmap.map.data.setStyle).toHaveBeenCalledTimes(1);
    unmount();
    expect(services.gmap.map.data.remove).toHaveBeenCalledTimes(1);
  });
  it('has play mode', () => {
    const store = createMockStore();
    const {unmount} = render(<GoogleMap map={testMap} mode="play" />, store);
    expect(services.gmap.map.setOptions.mock.calls[0]).toMatchSnapshot(
      'play settings'
    );
    expect(services.gmarkers.items).toHaveLength(1);
    expect(services.gmarkers.items[0].setMap).toHaveBeenCalledTimes(1);
    expect(services.gmarkers.items[0].setDraggable).toHaveBeenCalledTimes(1);
    expect(events.length).toBe(2);
    expect(events[1].event).toBe('click');
    events[1].func();
    expect(services.gmarkers.items[0].setPosition).toHaveBeenCalledTimes(1);
    events[1].func();
    expect(services.gmarkers.items[0].setPosition).toHaveBeenCalledTimes(2);
    expect(store.getState().position.selectedPosition).toMatchSnapshot(
      'update selected position'
    );
    unmount();
    expect(services.gmarkers.items).toHaveLength(0);
  });
  it('has result mode', () => {
    const props: GoogleMapProps = {
      map: testMap,
      mode: 'result',
      results: [
        {name: 'a', selected: {lat: 1, lng: 1}},
        {name: 'b', selected: {lat: 1, lng: 1}},
      ],
      initialPosition: {lat: 1, lng: 1},
    };
    const {unmount} = render(<GoogleMap {...props} />);
    expect(services.gmap.map.setOptions.mock.calls[0]).toMatchSnapshot(
      'result settings'
    );
    expect(services.gmarkers.items).toHaveLength(3);
    expect(services.gpolylines.items).toHaveLength(2);
    unmount();
    expect(services.gmarkers.items).toHaveLength(0);
    expect(services.gpolylines.items).toHaveLength(0);
  });
});

describe('Google, Street view', () => {
  it('renders and has containers in document', () => {
    render(<GoogleStreetView />);
    expect(screen.getByTestId('__GSTV__CONTAINER__')).toBeInTheDocument();
    expect(screen.getByTestId('__GSTV__')).toBeInTheDocument();
    expect(screen.getByTestId('__GSTV__')).toHaveStyle('height:100%');
  });
  it('has game mode', () => {
    render(<GoogleStreetView />);
    expect(services.gsv.sv.setPano).toHaveBeenCalledTimes(1);
    expect(services.gsv.sv.setOptions.mock.calls[0][0]).toMatchSnapshot(
      'options default'
    );
  });
  it('has static review mode', () => {
    render(<GoogleStreetView staticPos />);
    expect(services.gsv.sv.setPano).toHaveBeenCalledTimes(1);
    expect(services.gsv.sv.setOptions.mock.calls[0][0]).toMatchSnapshot(
      'options static'
    );
  });
});

describe('Google, custom marker', () => {
  it('defines a marker color for each player', () => {
    expect(config.maxPlayers).toBeLessThanOrEqual(svgMarkerColors.length);
  });
});
