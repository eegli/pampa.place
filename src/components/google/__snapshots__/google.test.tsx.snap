// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google, Map has play mode: play settings 1`] = `
Array [
  Object {
    "clickableIcons": false,
    "disableDefaultUI": true,
    "draggableCursor": "crosshair",
    "gestureHandling": "auto",
    "mapTypeControl": true,
    "mapTypeControlOptions": Object {
      "mapTypeIds": Array [
        "hybrid",
        "roadmap",
      ],
    },
    "mapTypeId": "roadmap",
  },
]
`;

exports[`Google, Map has play mode: update selected position 1`] = `
Object {
  "lat": 8,
  "lng": 8,
}
`;

exports[`Google, Map has preview mode: preview settings 1`] = `
Array [
  Object {
    "clickableIcons": false,
    "disableDefaultUI": true,
    "draggableCursor": "crosshair",
    "gestureHandling": "none",
    "mapTypeControl": false,
    "mapTypeControlOptions": Object {
      "mapTypeIds": Array [
        "hybrid",
        "roadmap",
      ],
    },
    "mapTypeId": "roadmap",
  },
]
`;

exports[`Google, Map has result mode: result settings 1`] = `
Array [
  Object {
    "clickableIcons": false,
    "disableDefaultUI": true,
    "draggableCursor": "crosshair",
    "gestureHandling": "auto",
    "mapTypeControl": true,
    "mapTypeControlOptions": Object {
      "mapTypeIds": Array [
        "hybrid",
        "roadmap",
      ],
    },
    "mapTypeId": "roadmap",
  },
]
`;

exports[`Google, Map renders and has containers in document 1`] = `<div />`;

exports[`Google, Map renders and has containers in document: unmount 1`] = `
Object {
  "calls": Array [
    Array [
      <div
        style="height: 100%;"
      />,
    ],
  ],
  "instances": Array [
    [Function],
  ],
  "invocationCallOrder": Array [
    2,
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": [Function],
    },
  ],
}
`;

exports[`Google, Street view has game mode: options default 1`] = `
Object {
  "addressControl": false,
  "clickToGo": true,
  "disableDoubleClickZoom": false,
  "fullscreenControl": false,
  "motionTracking": false,
  "motionTrackingControl": false,
  "pov": Object {
    "heading": 35,
    "pitch": 10,
  },
  "showRoadLabels": false,
  "zoom": 0,
  "zoomControl": false,
}
`;

exports[`Google, Street view has static review mode: options static 1`] = `
Object {
  "addressControl": false,
  "clickToGo": false,
  "disableDefaultUI": true,
  "disableDoubleClickZoom": true,
  "fullscreenControl": false,
  "motionTracking": false,
  "motionTrackingControl": false,
  "pov": Object {
    "heading": 35,
    "pitch": 10,
  },
  "showRoadLabels": false,
  "zoom": 0,
  "zoomControl": false,
}
`;
